gcloud-sdk/src/token_source/ext_creds_source/mod.rs

```diff
@@ -139,3 +175,187 @@ fn subject_token_from_json(
+                .into_iter()
+                .flat_map(|(k, v)| {
+                    v.to_str()
+                        .ok()
+                        .map(|v| AWSRequestHeader::new(k.to_string(), v.to_string()))
+                })
+                .collect(),
+        };
+        let payload =
+            serde_json::to_string(&payload).map_err(|e| Error::from(ErrorKind::TokenJson(e)))?;
+        let sts_token = utf8_percent_encode(&payload, NON_ALPHANUMERIC).to_string();
+
+        Ok(sts_token.into())
+    }
+
+    pub async fn get_aws_props() -> crate::error::Result<(Credentials, Region)> {
+        let imds_client = aws_config::imds::Client::builder().build();
+        let imds_credentials_provider =
+            aws_config::imds::credentials::ImdsCredentialsProvider::builder()
+                .imds_client(imds_client.clone())
+                .build();
+        let imds_region_provider = aws_config::imds::region::Builder::default()
+            .imds_client(imds_client)
+            .build();
+
+        let config = aws_config::load_from_env().await;
+        let region = match config.region() {
+            Some(region) => Some(region.to_owned()),
+            None => imds_region_provider.region().await,
+        }
+        .ok_or_else(|| {
+            Error::from(ErrorKind::ExternalCredsSourceError(
+                "region not found".to_string(),
+            ))
+        })?;
+        let provide_credentials = {
+            match config.credentials_provider() {
+                Some(provider) => provider.provide_credentials().await,
+                None => imds_credentials_provider.provide_credentials().await,
+            }
+        };
+        let credentials: Credentials = provide_credentials
+            .map_err(|e| Error::from(ErrorKind::ExternalCredsSourceError(e.to_string())))?
+            .into();
+        Ok((credentials, region))
+    }
+
+    #[derive(Debug, Serialize)]
+    struct AWSRequest {
+        url: String,
+        method: String,
+        headers: Vec<AWSRequestHeader>,
+    }
+
+    #[derive(Debug, Serialize, Clone)]
+    struct AWSRequestHeader {
+        key: String,
+        value: String,
+    }
+
+    impl AWSRequestHeader {
+        pub fn new(key: impl Into<String>, value: impl Into<String>) -> Self {
+            Self {
+                key: key.into(),
+                value: value.into(),
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use std::time::SystemTime;
+
+        use aws_config::Region;
+        use aws_credential_types::Credentials;
+        use chrono::NaiveDateTime;
+
+        use super::subject_token_aws;
+        #[tokio::test]
+        async fn sanity_check_subject_token() {
+            // This test uses the following implementation for reference
+            // https://github.com/yoshidan/google-cloud-rust/blob/8d09d6156dfb29965cd20539375896f16b3f739d/foundation/auth/src/token_source/external_account_source/aws_subject_token_source.rs#L381
```

Comment:
This test case is taken from https://github.com/yoshidan/google-cloud-rust/blob/8d09d6156dfb29965cd20539375896f16b3f739d/foundation/auth/src/token_source/external_account_source/aws_subject_token_source.rs#L381 with some modification because the original uses `ap-northeast-1b` instead of `ap-northeast-1` by mistake. GCP AIP 4117 implies that zone suffix should be removed.